src/data_readers/data_reader_hdf5_legacy.cpp

```diff
@@ -328,10 +328,13 @@ template <typename TensorDataType>
 bool hdf5_reader<TensorDataType>::fetch_datum(Mat& X, int data_id, int mb_idx) {
   prof_region_begin("fetch_datum", prof_colors[0], false);
 
-  assert_eq(sizeof(DataType)%sizeof(TensorDataType), 0);
-  assert_eq(X.Height(),
-            m_num_features / dc::get_number_of_io_partitions()
-            / (sizeof(DataType) / sizeof(TensorDataType)));
+  assert_eq(sizeof(DataType) % sizeof(TensorDataType), 0ul);
+  assert_eq((unsigned long) X.Height(),
+            m_num_features / dc::get_number_of_io_partitions());
+  //        / (sizeof(DataType) / sizeof(TensorDataType)));
+  // FIXME (trb 02/03/2023) -- when DataType is 'float' and
+  // TensorDataType is 'double', this last operand will be 0, which is
+  // invalid.
```

Comment:
Is this assert still valid with the last bit commented out?

Reply:
My thinking was that when `DataType` is `float` and `TensorDataType` is `float`, the commented-out term is 1, so that's the same. Then when `DataType` is `float` and `TensorDataType` is `double, the commented-out term is 0 and everything explodes. My decision to comment it out and leave the `FIXME` is due to the fact that `TensorDataType` should only actually be `float` in real LBANN usage as it stands today. I'll need to actually look at the surrounding code to see if it actually makes sense to leave it this way long-term.

Reply:
Why don't we add an explicit assert that `sizeof(DataType) == sizeof(TensorDataType)` so that this weird case is actually caught?

Reply:
Please have a look at my revised changes and the associated note. In particular, it's good to observe that `sizeof(DataType) % sizeof(TensorDataType)` is actually `constexpr`, so the first `assert_eq` can be replaced by a `constexpr` assertion. This seems to have been sufficient for the compiler to determine that the potentially-dangerous `<num> / (sizeof(DataType) / sizeof(TensorDataType))` operation could never be reached in a case in which `sizeof(DataType) / sizeof(TensorDataType) == 0` (as one would have hoped).